doc/org-roam.org

```diff
@@ -1380,6 +1201,141 @@ NOTE: Since ripgrep cannot read encrypted files, this function cannot find
 unlinked references within encrypted files.
 #+end_quote
 
+* Developer's Guide to Org-roam
+** Org-roam's Design Principle
+
+Org-roam is primarily motivated by the need for a dual representation. We
+(humans) love operating in a plain-text environment. The syntax rules of
+Org-mode are simple and fit snugly within our brain. This also allows us to use
+the tools and packages we love to explore and edit our notes. Org-mode is simply
+the most powerful plain-text format available, with support for images, LaTeX,
+TODO planning and much more.
+
+But this plain-text format is simply ill-suited for exploration of these notes:
+plain-text is simply not amenable for answering large-scale, complex queries
+(e.g. how many tasks do I have that are due by next week?). Interfaces such as
+Org-agenda slow to a crawl when the number of files becomes unwieldy, which can
+quickly become the case.
+
+At its core, Org-roam provides a database abstraction layer, providing a dual
+representation of what's already available in plain-text. This allows us
+(humans) to continue working with plain-text, while programs can utilize the
+database layer to perform complex queries. These capabilities include, but are
+not limited to:
+
+- link graph traversal and visualization
+- Instantaneous SQL-like queries on headlines
+  - What are my TODOs, scheduled for X, or due by Y?
+
+All of these functionality is powered by this database abstraction layer. Hence,
+at its core Org-roam's primary goal is to provide a resilient dual
+representation that is cheap to maintain, easy to understand, and is as
+up-to-date as it possibly can. Org-roam also then exposes an API to this
+database abstraction layer for users who would like to perform programmatic
+queries on their Org files.
+
+** Building Extensions and Advanced Customization of Org-roam
+
+Because Org-roam's core functionality is small, it is possible and sometimes
+desirable to build extensions on top of it. These extensions may one or more of
+the following functionalities:
+
+- Access to Org-roam's database
+- Usage/modification of Org-roam's interactive commands
+
+Org-roam provides no guarantees that extensions will continue to function as
+Org-roam evolves, but by following these simple rules, extensions can be made
+robust to local changes in Org-roam.
+
+1. Extensions should not modify the database schema. Any extension that requires
```

Comment:
I think that not all extensions make sense in Org-roam, because some applications have too specific needs. For one of my applications I've simply mimicked DB flow from Org-roam with custom scheme and extraction. And I decided not to go the length of building a generic tool for building custom schemes until I have another use case. And now I think that Org-roam could provide such functionality. E.g. as a user I simply provide a scheme and extractor function (that say takes an `id` and is executed in a buffer visiting note and returns a plist that needs to be inserted into db according to scheme) and Org-roam does all the lifting and maintenance. It should also provide versioning utilities via either migration function provided by user or full rebuild. This is something I've been thinking to implement, but now I wonder if there is interest in having it in Org-roam.

Reply:
Note: everything in the properties drawer is cached, and you can define your own accessor using `cl-defmethod` and reading the cached properties: `(org-roam-node-properties node)`, which should fulfill most users needs (e.g. having a last modified property for each node)

Reply:
I am bad at explaining things :) So I simply mean that sometimes you need custom database for subset of your notes (or all notes). And if it contains data that is not cached by org-roam, the only solution is to mimic org-roam solution for db. Maybe it's a very rare case and only I faced it. In any case, it is irrelevant to your doc update :)